Respect overriding the TEST_HOST build setting

gustn3965 opened the pull request:
…, even when using xcconfig.

Resolves <https://github.com/tuist/tuist/issues/6888>

### Short description 📝

 Fix Tuist overrides the `TEST_HOST` build setting for UnitTest targets, even when using xcconfig.


**Concerns**:
> If this issue fix is implemented, one concern is that users who have set defaultSettings to `.none` and have not explicitly provided a `TEST_HOST` in the code will face build failures in their projects.

> Personally, considering the implications of the comment regarding `defaultSettings`, it seems inconsistent to have code that overrides `TEST_HOST` with specific properties of the target, even when `defaultSettings` is set to `.none`. 
If it is `.none`, I believe users should have the right to customize their settings fully in accordance with that meaning.

>(I have personally experienced instances where I forgot the values I had previously set for properties despite using `.none`, resulting in additional debugging time.)

### How to test the changes locally 🧐


1. Write test code in ConfigGeneratorTests to ensure that when defaultSettings is either .`none` or `.recommended` and `.essential`, if `TEST_HOST` is included in `excluding`, then `TEST_HOST` should be nil.
2.  Generate the attached project with Tuist and check the `TEST_HOST` in the build settings of the unit test target (xcconfig attached).
[unittest.zip](https://github.com/user-attachments/files/17422586/unittest.zip)


### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

fortmarek commented on Sources/TuistGenerator/Generator/ConfigGenerator.swift:
```diff
@@ -321,6 +321,16 @@ final class ConfigGenerator: ConfigGenerating {
         var settings: SettingsDictionary = [:]
         settings["TEST_TARGET_NAME"] = .string("\(app.target.name)")
         if target.product == .unitTests {
+            if let defaultSettings = target.settings?.defaultSettings {
+                switch defaultSettings {
+                case let .essential(excludedKeys), let .recommended(excludedKeys):
+                    if excludedKeys.contains("TEST_HOST") {
+                        return [:]
+                    }
+                case .none:
+                    return [:]
+                }
+            }
             settings["TEST_HOST"] =
                 .string(
                     "$(BUILT_PRODUCTS_DIR)/\(app.target.productNameWithExtension)/$(BUNDLE_EXECUTABLE_FOLDER_PATH)/\(app.target.productName)"
```
I'd move the whole `testBundleTargetDerivedSettings` method into `DefaultSettingsProvider`. In the end, this settings is a sort of default settings and I don't think it belongs here. The implementation can look similar to `projectOverridableTargetDefaultSettings` – but you will need to update the `DefaultSettingsProvider` to also accept a `GraphTraverser`.

fortmarek reviewed (commented):
Thanks a lot for the PR! I think we can make the code a bit more readable by moving the method that sets the `TEST_HOST` to `DefaultSettingsProvider`.

gustn3965 commented:
@fortmarek  Thank you very much for the valuable review. Following your advice, I moved the `testBundleTargetDerivedSettings` method to `DefaultSettingsProvider`. `DefaultSettingsProvider` now requires a `GraphTraverser`, so I added a parameter to receive it from `ConfigGenerator`. Consequently, I also updated the `DefaultSettingsProviderTests` code.

fortmarek commented on Tests/TuistGeneratorTests/Generator/ConfigGeneratorTests.swift:
```diff
@@ -997,6 +997,44 @@ final class ConfigGeneratorTests: TuistUnitTestCase {
     }
 
+    func test_generateTargetConfig_when_defaultSettingsIsRecommendedWithExcludingTEST_HOST_then_TEST_HOSTIsNil() async throws {
+        let settings = Settings.test(defaultSettings: .essential(excluding: ["TEST_HOST"]))
+        try await generateTestTargetConfigTestHostSettings(settings)
+        let targetSettingsResult = try pbxTarget
+            .buildConfigurationList?
+            .buildConfigurations
+            .first { $0.name == "Debug" }?
+            .buildSettings
+            .toSettings()["TEST_HOST"]
+        XCTAssertEqual(targetSettingsResult, nil)
```
Can we follow the `// Given // When // Then` convention? You can take inspiration from the other test methods in this test suite 🤞

fortmarek reviewed (commented):
Two minor test-related nits, but the implementation looks good now 🙌

gustn3965 commented on Tests/TuistGeneratorTests/Settings/DefaultSettingsProviderTests.swift:
```diff
@@ -969,12 +994,19 @@ final class DefaultSettingsProvider_iOSTests: TuistUnitTestCase {
         XCTAssertEqual(got, multiplatformFrameworkTargetEssentialDebugSettings)
     }
+
+    private func graphTraverser(project: Project) -> GraphTraverser {
+        let graph = Graph.test(path: project.path)
+        let graphTraverser = GraphTraverser(graph: graph)
+        return graphTraverser
+    }
```
I think that’s a good point. I usually believe that repetitive code should be reduced. 
As you mentioned, it’s true that it doesn’t reduce a lot of code. It seems better to give each test case an independent meaning.

gustn3965 commented:
Thanks to @fortmarek , I was able to gain a new perspective instead of thinking that duplicated code must always be reduced.
I’m really enjoying it because of your review.

fortmarek reviewed (approved):
Thanks again for this PR 🙌 An amazing first contribution 👏

fortmarek commented:
@all-contributors add @gustn3965 for code
